packages/inject/src/inject.ts

```diff
@@ -0,0 +1,213 @@
+import { Config, WebchatEvent, WebchatEventType } from '@botpress/webchat'
+
+require('./inject.css')
+
+interface WebchatRef {
+  postMessage: (message: any, targetOrigin: string) => void
+  eventListener: {
+    handler: (event: WebchatEvent) => void
+    topics: WebchatEventHandlerTopics
+  }
+}
+
+type WebchatEventHandler = (evnt: WebchatEvent) => void
+type WebchatEventHandlerTopics = WebchatEventType[] | ['*']
+
+// full backward compatibility
+const DEFAULT_CHAT_ID = 'bp-web-widget'
+const DEFAULT_IFRAME_ID = 'bp-widget'
+
+const CHAT_REFS: { [chatId: string]: WebchatRef } = {}
+
+function _getContainerId(chatId?: string) {
+  return chatId ? `${chatId}-container` : DEFAULT_CHAT_ID
+}
+
+function _getIframeId(chatId: string) {
+  return chatId || DEFAULT_IFRAME_ID
+}
+
+function _injectDOMElement(
+  tagName: string,
+  selector: string,
+  options: { [key: string]: string } = {}
+): HTMLElement | void {
+  const element = document.createElement(tagName)
+  // @ts-ignore
+  Object.entries(options).forEach(([attrName, attrValue]) => (element[attrName] = attrValue))
```

Comment:
**Suggestion:** if you ever think this adds value or could impress someone :sweat_smile: :

```ts
function _injectDOMElement<T extends keyof HTMLElementTagNameMap>(
  tagName: T,
  selector: string,
  options: {
    [key in keyof Partial<HTMLElementTagNameMap[T]>]?: HTMLElementTagNameMap[T][Extract<
      keyof HTMLElementTagNameMap[T],
      string
    >]
  } = {}
): HTMLElement | void {
  const element = document.createElement(tagName)

  for (const key in options) {
    element[key] = options[key]!
  }
// [...]
```

Reply:
It seems to me that it's a lot playing around with ts, that makes the code less readable and less easy to maintain.

Reply:
It might impress some people tho ;)